exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/RuntimeFilterRecordBatch.java

```diff
@@ -0,0 +1,222 @@
+    container.transferIn(incoming.getContainer());
+    originalRecordCount = incoming.getRecordCount();
+    sv2.setOriginalRecordCount(originalRecordCount);
+    try {
+      applyRuntimeFilter();
+    } catch (SchemaChangeException e) {
+      throw new UnsupportedOperationException(e);
+    }
+    return getFinalOutcome(false);
+  }
+
+  @Override
+  public void close() {
+    if (sv2 != null) {
+      sv2.clear();
+    }
+    super.close();
+  }
+
+  @Override
+  protected boolean setupNewSchema() throws SchemaChangeException {
+    if (sv2 != null) {
+      sv2.clear();
+    }
+
+    switch (incoming.getSchema().getSelectionVectorMode()) {
+      case NONE:
+        if (sv2 == null) {
+          sv2 = new SelectionVector2(oContext.getAllocator());
+        }
+        break;
+      case TWO_BYTE:
+        sv2 = new SelectionVector2(oContext.getAllocator());
+        break;
+      case FOUR_BYTE:
+
+      default:
+        throw new UnsupportedOperationException();
+    }
+
+    if (container.isSchemaChanged()) {
+      container.buildSchema(SelectionVectorMode.TWO_BYTE);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   *
+   * @return True means rows are filtered by the RuntimeFilter , SelectionVector2 is set.
+   * False means not affected by the RuntimeFilter, SelectionVector2 is no set.
+   * @throws SchemaChangeException
+   */
+  private void applyRuntimeFilter() throws SchemaChangeException {
+    RuntimeFilterWritable runtimeFilterWritable = context.getRuntimeFilter();
+    if (runtimeFilterWritable == null) {
+      sv2.setRecordCount(incoming.getRecordCount());
+      return;
+    }
+    if (originalRecordCount <= 0) {
+      sv2.setRecordCount(0);
+      return ;
+    }
+    List<BloomFilter> bloomFilters = runtimeFilterWritable.unwrap();
+    if (hash64 == null) {
+      ValueVectorHashHelper hashHelper = new ValueVectorHashHelper(this, context);
+      try {
+        //generate hash helper
+        this.toFilterFields = runtimeFilterWritable.getRuntimeFilterBDef().getProbeFieldsList();
+        List<LogicalExpression> hashFieldExps = new ArrayList<>();
+        List<TypedFieldId> typedFieldIds = new ArrayList<>();
+        for (String toFilterField : toFilterFields) {
+          SchemaPath schemaPath = new SchemaPath(new PathSegment.NameSegment(toFilterField), ExpressionPosition.UNKNOWN);
+          TypedFieldId typedFieldId = container.getValueVectorId(schemaPath);
+          this.field2id.put(toFilterField, typedFieldId.getFieldIds()[0]);
+          typedFieldIds.add(typedFieldId);
+          ValueVectorReadExpression toHashFieldExp = new ValueVectorReadExpression(typedFieldId);
+          hashFieldExps.add(toHashFieldExp);
+        }
+        hash64 = hashHelper.getHash64(hashFieldExps.toArray(new LogicalExpression[hashFieldExps.size()]), typedFieldIds.toArray(new TypedFieldId[typedFieldIds.size()]));
+      } catch (Exception e) {
+        throw UserException.internalError(e).build(logger);
+      }
+    }
```

Comment:
above setup of `Hash64` should not happen for each `incoming` record batch, since that will be very expensive. It should happen only once during `setupNewSchema` phase until next `OK_NEW_SCHEMA`.

Reply:
When the `setupNewSchema` was called, the `RuntimeFilterWritable` may have not been received . So the `hash64` should be initialized at the `doWork`. Since there's `hash64 == null` checking and initializing `hash64` logic, this part of code will only run one time.

Reply:
make sense, but could you explain at which case ,to a query , which will have consecutive incoming `RecordBatch` but some with different schema? This really confuse me and make the solving logic complex.

Reply:
a simple example will be Scan is reading 2 json files in 2 different record batches. Both files have a column of type map lets say m. File1 have children in map a,b of type int whereas file2 have children a,b,c. In this case scan will generate OK_NEW_SCHEMA